cpp/include/cudf/fixed_point/floating_conversion.hpp

```diff
@@ -609,6 +640,641 @@ CUDF_HOST_DEVICE inline constexpr T divide_power10(T value, int exp10)
   }
 }
 
+/**
+ * @brief Perform a bit-shift left, guarding against undefined behavior
+ *
+ * @tparam IntegerType Type of input unsigned integer value
+ * @param value The integer whose bits are being shifted
+ * @param bit_shift The number of bits to shift left
+ * @return The bit-shifted integer, except max value if overflow would occur
+ */
+template <typename IntegerType, CUDF_ENABLE_IF(cuda::std::is_unsigned_v<IntegerType>)>
+CUDF_HOST_DEVICE inline IntegerType guarded_left_shift(IntegerType value, int bit_shift)
+{
+  // Bit shifts larger than this are undefined behavior
+  static constexpr int max_safe_bit_shift = cuda::std::numeric_limits<IntegerType>::digits - 1;
```

Comment:
Well constexpr by itself doesn't necessarily guarantee that it will run at compile time, it only states that it can (no `consteval` yet). Making it `static` makes sure that it only happens once no matter what.